presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/LimitPushDown.java

```diff
@@ -117,13 +125,43 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<LimitConte
         {
             LimitContext limit = context.get();
 
-            if (limit != null &&
-                    node.getAggregations().isEmpty() &&
+            if (limit != null) {
+                checkArgument(!node.getSampleWeight().isPresent(), "DISTINCT aggregation has sample weight symbol");
+                if (node.getAggregations().isEmpty() &&
                     node.getOutputSymbols().size() == node.getGroupBy().size() &&
                     node.getOutputSymbols().containsAll(node.getGroupBy())) {
-                checkArgument(!node.getSampleWeight().isPresent(), "DISTINCT aggregation has sample weight symbol");
-                PlanNode rewrittenSource = context.rewrite(node.getSource());
-                return new DistinctLimitNode(idAllocator.getNextId(), rewrittenSource, limit.getCount(), Optional.empty());
+                    PlanNode rewrittenSource = context.rewrite(node.getSource());
+                    return new DistinctLimitNode(idAllocator.getNextId(), rewrittenSource, limit.getCount(), Optional.empty(), node.getGroupBy(), Optional.empty(), Optional.empty());
+                }
+                else if (node.getOutputSymbols().size() == node.getGroupBy().size() + node.getAggregations().size() &&
```

Comment:
i think it would be a sufficient check here to just see if all the functions are "arbitrary".